python/lsst/cell_coadds/_fits.py

```diff
@@ -0,0 +1,407 @@
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = (
+    "CellCoaddFitsFormatter",
+    "CellCoaddFitsReader",
+    "writeExplodedCoaddAsFits",
+    "writeMultipleCellCoaddAsFits",
+    "writeStitchedCoaddAsFits",
+)
+
+import os
+from typing import TYPE_CHECKING, Any, Mapping
+
+import lsst.afw.geom as afwGeom
+import lsst.afw.image as afwImage
+import numpy as np
+from astropy.io import fits  # type: ignore[import]
+from lsst.afw.image import ImageD, ImageF
+from lsst.daf.base import PropertySet
+from lsst.geom import Box2I, Extent2I, Point2I
+from lsst.obs.base.formatters.fitsGeneric import FitsGenericFormatter
+from lsst.skymap import Index2D
+
+from ._common_components import CoaddUnits, CommonComponents
+from ._identifiers import CellIdentifiers, PatchIdentifiers
+from ._image_planes import OwnedImagePlanes
+from ._multiple_cell_coadd import MultipleCellCoadd, SingleCellCoadd
+from ._uniform_grid import UniformGrid
+
+if TYPE_CHECKING:
+    from ._exploded_coadd import ExplodedCoadd  # pragma: no cover
+    from ._stitched_coadd import StitchedCoadd  # pragma: no cover
+
+
+class CellCoaddFitsFormatter(FitsGenericFormatter):
+    """Interface for writing and reading cell coadds to/from FITS files.
+
+    This assumes the existence of readFits and writeFits methods (for now).
+    """
+
+
+class CellCoaddFitsReader:
+    """A reader class to read from a FITS file and produce cell-based coadds.
+
+    This reader class has read methods that can either return a single
+    component without reading the entire file (e.g., readBBox, readWcs)
+    and read methods that return a full coadd (e.g.,
+    readAsMultipleCellCoadd, readAsExplodedCellCoadd, readAsStitchedCoadd).
+
+    Parameters
+    ----------
+    filename : `str`
+        The name of the FITS file to read.
+    """
+
+    def __init__(self, filename: str) -> None:
+        if not os.path.exists(filename):
+            raise FileNotFoundError(f"File {filename} not found")
+
+        self.filename = filename
+
+    def readAsMultipleCellCoadd(self) -> MultipleCellCoadd:
+        """Read the FITS file as a MultipleCellCoadd object."""
+        with fits.open(self.filename) as hdu_list:
+            data = hdu_list[1].data
+
+            # Read in WCS
+            ps = PropertySet()
+            ps.update(hdu_list[0].header)
+            wcs = afwGeom.makeSkyWcs(ps)
+
+            common = CommonComponents(
+                units=CoaddUnits(1),  # TODO: read from FITS TUNIT1
+                wcs=wcs,
+                band=hdu_list[1].header["BAND"],
+                identifiers=PatchIdentifiers(
+                    skymap=hdu_list[1].header["SKYMAP"],
+                    tract=hdu_list[1].header["TRACT"],
+                    patch=Index2D(x=hdu_list[1].header["PATCH_X"], y=hdu_list[1].header["PATCH_Y"]),
+                    band=hdu_list[1].header["BAND"],
```

Comment:
This is mostly pre-existing, but seeing this makes me think we probably want to also include the sequential patch and cell IDs in the identifiers structs, save the sequential patch index here, and save the sequential cell index in a new column.  Being able to get most of the data ID but not _quite_ the patch ID from the file would be a bit frustrating.

Reply:
[DM-39191](https://jira.lsstcorp.org/browse/DM-39191)